Fixed point smooth linear advance (makes it available to all 32 bit boards)

dbuezas opened the pull request:
Port smooth linear advance to fixed point arithmetic.
Followup to #27710

~~There is still one float calculation inside the ISR~~

~~`        target_adv_steps_q13 = lookahead(t) * (Planner::extruder_advance_K[0]*(1UL<<13));// todo keep k in q format`~~

~~I'll later put extruder_advance_k behind a getter setter where also a fixed point version in q18.13 format will be kept up to date for use inside the isr.~~

~~Opening the PR early because it may be already fast enough for tests in boards without an FPU.~~

Also changed to more conservative initial values for tau (20ms instead of 10) and smooth advance isr frequency (1kHz instead of 5kHz)

Smooth linear advance should now work in any 32 bit board and maybe (__just maybe__) in an 8 bit one.

dbuezas committed:
almost done. seems to work

(cherry picked from commit a06e1a324f9c9a347085eb04e4eb15e1803c2f45)

dbuezas committed:
conservative settings

dbuezas commented:
Note I'm using different positions for the fixed point in different variables to maximize decimal resolution depending on the magnitude of the numbers that they store.
The position of the point is noted in the variable names with a posfix. E.g `xxx_q30` means xxx<<30.

dbuezas commented:
If necessary, it could be optimized a bit more by:
* Experimentally finding the lowest isr frequency that doesn't affect quality 
* Tweaking the point positions so the int64 cast isn't necessary when multiplying (and confirm the loss of resolution doesn't affect quality)
* Getting rid of rounding in fixed point multiplications

But that would require a lot of care to avoid overflows and I think it wont really make such a big difference, just a couple of cycles per multiplication.

rq3 commented:
Printing as I type on an Anycubic Predator delta with STM32F103 (no FPU).
I see no issues.

dbuezas committed:
remove the last float

(cherry picked from commit 6e1e42c144f73d5bd1370c8d9a91733c1f4764f6)

# Conflicts:
#	Marlin/src/gcode/feature/advance/M900.cpp
#	Marlin/src/module/planner.cpp
#	Marlin/src/module/planner.h

dbuezas committed:
formatting

dbuezas committed:
fix stepper.get_advance_k => planner.get_advance_k

dbuezas committed:
stepper.set_advance_k => planner.set_advance_k

dbuezas committed:
fix menu with multiple extruders

dbuezas commented:
I'm doing some test prints to compare floating point vs fixed point branches. There is a slight (but visible) loss in quality in fixed point. This may be the lower time resolution of the less frequent ISR. May have to increase it back.

dbuezas committed:
format

(cherry picked from commit 66d41fffd8a2b2726f6635098dca357c221e3916)

classicrocker883 commented on Marlin/src/lcd/menu/menu_tune.cpp:
```diff
@@ -216,10 +216,13 @@ void menu_tune() {
-        EDIT_ITEM_N(float42_52, e, MSG_ADVANCE_K_E, &planner.extruder_advance_K[e], 0, 10);
+      EXTRUDER_LOOP() {
+        editable.decimal = planner.get_advance_k(e);
+        EDIT_ITEM_N(float42_52, e, MSG_ADVANCE_K_E, &editable.decimal, 0.0f, 10.0f, []{ planner.set_advance_k(editable.decimal, e); });
```
try `e` in the brackets `[]`

```y
EDIT_ITEM_N(float42_52, e, MSG_ADVANCE_K_E, &editable.decimal, 0.0f, 10.0f, [e]{ planner.set_advance_k(editable.decimal, e); });
```

and in **menu_advanced.cpp** line 125 / 761

dbuezas replied:
Good catch. Other places use MenuItemBase::itemIndex instead. I prefer capturing e as you suggest but I can't test multiple extruders so i'll just follow the existing pattern

dbuezas committed:
uint32_t get_advance_k_q15

(cherry picked from commit fb324b5d8f2f74c256afbadfe099797785094bef)

dbuezas committed:
minor cleanup

(cherry picked from commit c2e360a2357201ee836f3738eddf88e39c6f75eb)

dbuezas committed:
fix overflow

(cherry picked from commit e706ec93efbfb555fd0e0922dcd881c0e95d6fcc)

dbuezas committed:
Even more conservative isr frequency

(cherry picked from commit 00403f5d8d19ad75272402f66c5df3beec3bb98d)

dbuezas commented:
Found the problem.

15 minute benchy:

Left to right: FLOAT ---- FIXED POINT BEFORE ----- FIXED POINT NOW (fixed!)

Float and new fixed point look identical, can't distinguish them.
![image](https://github.com/user-attachments/assets/8615a15e-021c-43d7-838e-19579733e953)

dbuezas committed:
fix menu_tune multiple extruders

dbuezas committed:
use active extruder

(cherry picked from commit ab7d10af0d71269f39222843f6ab48e45b7f704f)

dbuezas committed:
fix dwin

dbuezas commented:
@rq3 Can you try again? the version you tried above had an overflow that worsened quality.

rq3 commented:
> @rq3 Can you try again? the version you tried above had an overflow that worsened quality.

Yes indeed. Just need to mod the files, recompile, and print. Give me a few hours.

classicrocker883 commented:
could have `SMOOTH_LIN_ADVANCE` enabled in the tests files **buildroot/tests/STM32F103RE_creality** when ProUI and JyersUI are enabled

dbuezas commented:
Glad to hear!
You can now increase jerk and acceleration (particularly for inner walls and infill) and the machine will now actually do what you ask for.

dbuezas commented:
Do you notice any quality improvement at equal printing time?

rq3 commented:
> Do you notice any quality improvement at equal printing time?

Comparing old and new Benchys printed identically (same g-code, speeds, temps, etc.) I note less ringing and stringing.
![Smooth_Benchy_2](https://github.com/user-attachments/assets/d84b2950-be05-4098-997f-d5cf3f836bb0)

dbuezas committed:
Merge branch 'bugfix-2.1.x' into dbuezas/fixed-point-smooth-linear-advance

dbuezas commented:
@rq3 excellent. Less stringing should come from the fact that smooth linear advance is aware of jerk too, and less ringing likely because of smoothing and e-synch. It should also be a tad faster (depending on what k and acceleration you used).
Any comment on seams and top surfaces?

I suggest you try a print with high acceleration: internal solid infills are particularly amazing to watch at high acceleration (now unlimited) and high jerk (now doesn't make the extruder scream for its life).

classicrocker883 commented on Marlin/src/module/stepper.cpp:
```diff
@@ -2904,21 +2909,21 @@ hal_timer_t Stepper::block_phase_isr() {
           delayBuffer.index = 0;
       }
 
-      xy_float_t lookback(shaping_time_t t /* in stepper timer ticks */) {
```
might be better to name `t` something like `stepper_ticks` and `past_i` maybe `buffer_index` just to be more descriptive

dbuezas committed:
avoid modulo

(cherry picked from commit 7f16f04be8ae04ec1fe77c5a111a936de7e32d77)

dbuezas committed:
better naming

Suggested by @classicrocker883

(cherry picked from commit e6f311de462c1d63eab38cee1f5ba803985ff463)

thinkyhead committed:
prettify

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/27818

thinkyhead committed:
Limit index in getter/setter

thinkyhead committed:
reset of other_extruder_advance_K

thinkyhead committed:
distinct other_extruder_advance_K

dbuezas commented on Marlin/src/module/settings.cpp:
```diff
@@ -3763,6 +3763,9 @@ void MarlinSettings::reset() {
     #else
       planner.set_advance_k(ADVANCE_K);
+      #if ENABLED(ADVANCE_K_EXTRA)
+        EXTRUDER_LOOP() other_extruder_advance_K[e] = ADVANCE_K;
```
Good catch. The setter should be called so the fixed point vars are also updated

dbuezas replied:
oh not here.

thinkyhead committed:
misc tweaks

thinkyhead committed:
misc cleanup, public access

dbuezas commented on Marlin/src/module/settings.cpp:
```diff
@@ -2643,24 +2643,13 @@ void MarlinSettings::postprocess() {
-            planner.set_advance_k(extruder_advance_K[0]);
-          #endif
-        }
+        if (!validating) COPY(planner.extruder_advance_K, extruder_advance_K);
```
If you just copy them, the fixed point variants won't be updated

dbuezas replied:
That's why I made it private and added setters and getters: to ensure nothing sets the float without setting the fixed point vars

thinkyhead committed:
notes

thinkyhead committed:
combined loop

thinkyhead committed:
relocate settings reset

thinkyhead committed:
clean

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/27818

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/27818

thinkyhead committed:
more ALIM

thinkyhead committed:
Try this "capture" thing

thinkyhead committed:
etc

thinkyhead committed:
no capture on action

thinkyhead committed:
_BV32(24) and _BV32(30)

thinkyhead committed:
ensure first if() block

dbuezas commented on Marlin/src/module/stepper.cpp:
```diff
@@ -2922,11 +2922,11 @@ hal_timer_t Stepper::block_phase_isr() {
-        if (delay_steps>= IS_COMPENSATION_BUFFER_SIZE) {
-          // this means the buffer is too small. TODO: how to inform user?
+        if (delay_steps >= IS_COMPENSATION_BUFFER_SIZE) {
+          // this means the buffer is too small. TODO: set a flag to inform user
```
I think we can remove both the check and the comment after all. If the buffer is not large enough it is because IS frequency was set lower than the minimum and the IS buffers are also too small. 
If that happens, the least of the problems is linear advance

thinkyhead committed:
use switch

thinkyhead committed:
check only if debugging

thinkyhead committed:
some review changes

thinkyhead committed:
final tweaks

thinkyhead committed:
encapsulate

dbuezas commented on Marlin/src/module/stepper.cpp:
```diff
@@ -2910,23 +2910,22 @@ hal_timer_t Stepper::block_phase_isr() {
         }
+        FORCE_INLINE xy_long_t past_item(const uint16_t n) {
+          const uint16_t i = index + IS_COMPENSATION_BUFFER_SIZE - n;
+          return buffer[i < IS_COMPENSATION_BUFFER_SIZE ? i : i - IS_COMPENSATION_BUFFER_SIZE];
```
I think we can save one operation here:

If (index >= n) return index - n; // 2 ops total
Else return  index + SIZE - n; // 3 ops total

Instead of 3 and 4

thinkyhead replied:
Subtraction is required to determine "greater than" state anyway, so…
```cpp
const int16_t i = int16_t(index) - n;
return buffer[i > 0 ? i : i + IS_COMPENSATION_BUFFER_SIZE];
```

thinkyhead committed:
skip extra add

dbuezas commented on Marlin/src/module/stepper.cpp:
```diff
@@ -2915,8 +2915,8 @@ hal_timer_t Stepper::block_phase_isr() {
-          const uint16_t i = index + IS_COMPENSATION_BUFFER_SIZE - n;
-          return buffer[i < IS_COMPENSATION_BUFFER_SIZE ? i : i - IS_COMPENSATION_BUFFER_SIZE];
+          const int16_t i = int16_t(index) - n;
+          return buffer[i > 0 ? i : i + IS_COMPENSATION_BUFFER_SIZE];
```
Being nitpicky here, but i think the int casting also costs an extra op, so this is still 3 and 4 ops

dbuezas replied:
Oh, maybe not. The int will always be bit equal to a uint since it is positive.

thinkyhead replied:
The casting is intended to prevent the compiler complaining about signed underflow but the assembly language to subtract 16 bits from 16 bits should not differ. Perhaps a `static_cast` will help to prevent any runtime sign-ish behavior.

dbuezas replied:
In that case, your version is the cleanest

thinkyhead committed:
oopsie

grizzleeadam commented:
I've been starting to re-calibrate my printer using this new mode and the results are pretty promising. However, it seems I can't change the input shaping frequency while printing - any attempt (whether via g-code or the LCD menu) causes the printer to freeze and reboot. Wondering if it has to do with the Input_Shaping_E_Synch option (which is enabled). Any way to g-code to disable that function without reflashing the board? (I use manual mesh leveling so it's a pain to recal every time I flash)

vovodroid commented:
@grizzleeadam 
> I use manual mesh leveling so it's a pain to recal every time I flash

You can output existing mesh with G29, and make g-code file creating mesh back also with G29.

By a way, what is link between input shaping calibration and LA? Aren't IS parameters depend on mechanical properties only?

grizzleeadam commented:
That’s helpful, thank you.

The Smooth_Lin_Adv feature includes a connection with input shaping, i
don’t understand it well but there is a line to enable/disable that
particular feature in configuration_adv.

I am recalibrating my input shaping after installing the new firmware just
to be thorough with all the new settings.

On Fri, May 16, 2025 at 11:20 PM Vovodroid ***@***.***> wrote:

> *vovodroid* left a comment (MarlinFirmware/Marlin#27818)
> <https://github.com/MarlinFirmware/Marlin/pull/27818#issuecomment-2888128735>
>
> @grizzleeadam <https://github.com/grizzleeadam>
>
> I use manual mesh leveling so it's a pain to recal every time I flash
>
> You can output existing mesh with G29, and make g-code file creating mesh
> back also with G29.
>
> By a way, what is link between input shaping calibration and LA? Aren't IS
> parameters depend on mechanical properties only?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/27818#issuecomment-2888128735>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A6SRTTWN4UNJYS76QQSXV7L263IJBAVCNFSM6AAAAAB35JKNZWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQOBYGEZDQNZTGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dbuezas commented:
I noticed that yesterday too. 
The link between Smooth linear advance and input shaping is the "e-synch" feature, which tracks how input shaping affects motion and makes extrusion follow that instead of the clean planned extrusion. It's inside smooth linear advance only because the new approach makes it easy to factor in.

Interestingly the crash doesn't happen if input shaping parameters are changed via gcode. I assume the difference is that the menu doesn't synchronise with the planner and the internal shaping params (like delay and factor) are changed while in use. That must be causing ludicrous extrusion speeds which make the firmware spend 100% of the time in the stepper interrupt, starving the watchdog and eventually resulting in halting. In short: a bug in e-synch

vovodroid commented:
> the link between Smooth linear advance and input shaping is the "e-synch"

But why recalibrate IS itself?

dbuezas commented:
Recalibration shouldn't be necessary

grizzleeadam commented:
> > the link between Smooth linear advance and input shaping is the "e-synch"
> 
> But why recalibrate IS itself?

I rushed my calibration when I did it the first time over a year ago (haven't tinkered much since), was just doing it as a reassurance - not because this commit needed it.



> I noticed that yesterday too. The link between Smooth linear advance and input shaping is the "e-synch" feature, which tracks how input shaping affects motion and makes extrusion follow that instead of the clean planned extrusion. It's inside smooth linear advance only because the new approach makes it easy to factor in.
> 
> Interestingly the crash doesn't happen if input shaping parameters are changed via gcode. I assume the difference is that the menu doesn't synchronise with the planner and the internal shaping params (like delay and factor) are changed while in use. That must be causing ludicrous extrusion speeds which make the firmware spend 100% of the time in the stepper interrupt, starving the watchdog and eventually resulting in halting. In short: a bug in e-synch

For me, the crash was happening with both gcode and with the menu. I tried adding a 100ms dwell before and after the gcode and it was still crashing at the first M593 command. I suppose I could try disabling the E sync and see if it still does it for me. I'm on a SKR E3 Mini v1.2, bit of an older setup.

dbuezas commented:
> I suppose I could try disabling the E sync and see if it still does it for me

That would be a good test, please post the result. Maybe gcode didn't fail for me is just because I made a frequency sweep from low to high. So the delay between echos got always shorter and the algorithm never had to look at an unset value of the lookback buffer.

dbuezas commented:
I think I know what the problem is.
The code assumes that the minimum frequency possible is the minimum set in the config for any axis. A lower frequency would mean a bigger buffer, and if one goes below that, then there's a buffer underflow and it starts reading echoes from memory where other vars store stuff. These values may happen to be large values, the firmware tries to extrude at ridiculously high speeds and eventually resets.

Since E-SYNCH doesn't check if shaping is enabled, and disabling it sets frequency to zero, it also creates the same problem, but worse: the delay is -1 in uint, so a very large value.
Interestingly, the input shaping buffers are also set to the minimum frequency, but the menu still allows you to go lower, which would also make the input shaper go crazy

dbuezas commented:
@tombrazier: by default SHAPING_MIN_FREQ is the minimum of the freqs set at build time, but the menu allows going below them. Doesn't that make the input shaper break?

dbuezas commented:
PR to fix that: https://github.com/MarlinFirmware/Marlin/pull/27862

grizzleeadam commented:
What about the manual override for the min frequency? I have that commented
out by default - should that be set to something?

On Sat, May 17, 2025 at 6:45 AM David Buezas ***@***.***>
wrote:

> *dbuezas* left a comment (MarlinFirmware/Marlin#27818)
> <https://github.com/MarlinFirmware/Marlin/pull/27818#issuecomment-2888404486>
>
> PR to fix that: #27862
> <https://github.com/MarlinFirmware/Marlin/pull/27862>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/27818#issuecomment-2888404486>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A6SRTTWO4NHRTRQUMZRLSYT2644QJAVCNFSM6AAAAAB35JKNZWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQOBYGQYDINBYGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dbuezas commented:
You can also set it, the issue is that Marlin allows you to then set a lower frequency and that.
Without the PR I linked above, it will still crash when IS is disabled in the UI though

tombrazier commented:
> @tombrazier: by default SHAPING_MIN_FREQ is the minimum of the freqs set at build time, but the menu allows going below them. Doesn't that make the input shaper break?

No, it just reduces the effectiveness of IS at higher speeds, as explained in the comment above the input shaping section of the config.

grizzleeadam commented:
I tried the following, one at a time, using latest 2.1.x-bugfix:

1) SMOOTH_LIN_ADV disabled: M593 to lower values works, even with manual SHAPING_MIN_FREQ not set.

2) SMOOTH_LIN_ADV enabled, default settings - M593 freezes.

3) Disabled INPUT_SHAPING_E_SYNCH and set SHAPING_MIN_FREQ & SHAPING_MAX_STEPRATE - M593 works. It seems all three of these must be set; any other combination causes M593 to freeze.

I'll repeat the same with your new PR linked above.

grizzleeadam commented:
With your PR 27862:

SMOOTH_LIN_ADV enabled, INPUT_SHAPING_E_SYNCH enabled;
SHAPING_MIN_FREQ & SHAPING_MAX_STEPRATE commented out;

M593 worked. However, the stepper motor started making some ugly sounds as the speed increased.

dbuezas commented:
> M593 worked. However, the stepper motor started making some ugly sounds as the speed increased.

Most likely you had set a frequency below SHAPING_MIN_FREQ. I'll continue the discussion in the open pr, thanks for the reports!

CRCinAU commented:
This looks to potentially have caused a new error in DWIN PROUI:
```
In file included from Marlin/src/lcd/e3v2/proui/dwin.cpp:36:
Marlin/src/lcd/e3v2/proui/dwin.cpp: In function 'void drawTuneMenu()':
Marlin/src/lcd/e3v2/proui/dwin.cpp:3549:104: error: lvalue required as unary '&' operand
 3549 |       EDIT_ITEM(ICON_MaxAccelerated, MSG_ADVANCE_K, onDrawPFloat3Menu, setLA_K, &planner.get_advance_k());
      |                                                                                                        ^
Marlin/src/lcd/e3v2/proui/menus.h:61:59: note: in definition of macro 'EDIT_ITEM'
   61 | #define EDIT_ITEM(I,L,V...) editItemAdd(I, GET_TEXT_F(L), V)
      |                                                           ^
Marlin/src/lcd/e3v2/proui/dwin.cpp: In function 'void drawMotionMenu()':
Marlin/src/lcd/e3v2/proui/dwin.cpp:3687:104: error: lvalue required as unary '&' operand
 3687 |       EDIT_ITEM(ICON_MaxAccelerated, MSG_ADVANCE_K, onDrawPFloat3Menu, setLA_K, &planner.get_advance_k());
      |                                                                                                        ^
Marlin/src/lcd/e3v2/proui/menus.h:61:59: note: in definition of macro 'EDIT_ITEM'
   61 | #define EDIT_ITEM(I,L,V...) editItemAdd(I, GET_TEXT_F(L), V)
      |  
```
